rust/azure_iot_operations_services/src/leased_lock/lease.rs

```diff
@@ -0,0 +1,326 @@
+                request_timeout,
+                None,
+                SetOptions {
+                    set_condition: SetCondition::OnlyIfEqualOrDoesNotExist,
+                    expires: Some(lease_expiration),
+                },
+            )
+            .await;
+
+        let state_store_response = a?;
+
+        if state_store_response.response {
+            self.current_fencing_token
+                .lock()
+                .expect("Could not lock mutex")
+                .clone_from(&state_store_response.version);
+
+            Ok(state_store_response.version.expect(
+                "Got None for fencing token. A lease without a fencing token is of no use.",
+            ))
+        } else {
+            *self
+                .current_fencing_token
+                .lock()
+                .expect("Could not lock mutex") = None;
+
+            Err(Error(ErrorKind::KeyAlreadyLeased))
+        }
+    }
+
+    /// Attempts to acquire a lease, returning if it cannot be acquired after one attempt.
+    ///
+    /// `lease_expiration` is how long the lease will remain held in the State Store after acquired, if not released before then.
+    /// `request_timeout` is the maximum time the function will wait for receiving a response from the State Store service, it is rounded up to the nearest second.
+    ///
+    /// Returns Ok with a fencing token (`HybridLogicalClock`) if completed successfully, or `Error(KeyAlreadyLeased)` if lease is not acquired.
+    /// # Errors
+    /// [`struct@Error`] of kind [`InvalidArgument`](ErrorKind::InvalidArgument) if the `request_timeout` is zero or > `u32::max`
+    ///
+    /// [`struct@Error`] of kind [`ServiceError`](ErrorKind::ServiceError) if the State Store returns an Error response
+    ///
+    /// [`struct@Error`] of kind [`UnexpectedPayload`](ErrorKind::UnexpectedPayload) if the State Store returns a response that isn't valid for a `Set` request
+    ///
+    /// [`struct@Error`] of kind [`AIOProtocolError`](ErrorKind::AIOProtocolError) if there are any underlying errors from the command invoker
+    ///
+    /// [`struct@Error`] of kind [`KeyAlreadyLeased`](ErrorKind::KeyAlreadyLeased) if the `lease` is already in use by another holder
+    /// # Panics
+    /// Possible panic if, for some error, the fencing token (a.k.a. `version`) of the acquired lease is None.
+    pub async fn acquire(
+        &mut self,
+        lease_expiration: Duration,
+        request_timeout: Duration,
+        renewal_period: Option<Duration>,
+    ) -> Result<HybridLogicalClock, Error> {
+        // Stop auto-renewal.
+        if !self.auto_renewal_cancellation_token.is_cancelled() {
+            self.auto_renewal_cancellation_token.cancel();
+        }
+
+        let acquire_result = self
+            .internal_acquire(lease_expiration, request_timeout)
+            .await;
+
+        if renewal_period.is_some() {
+            self.auto_renewal_cancellation_token = CancellationToken::new();
```

Comment:
`auto_renewal_cancellation_token` might need to be under a Mutex on the Client struct in this case, otherwise a clone of the Client won't necessarily get the new CancellationToken I think? I'm not sure this is an issue within your implementation since you always cancel the current one and create a new one before cloning the Client (so there's never a cloned client that lasts longer than the lifetime of one token), but if the customer clones the Lease, I think things might be able to get a little funky?

Reply:
If this changes, then I think that will also remove the need for &mut self in this fn signature

Reply:
Change CancellationToken for Notify in lease client. Also removed &mut in argument